daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Normalization.scala

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package transaction
+
+import com.daml.lf.value.{Value => V}
+
+import com.daml.lf.transaction.Node.{
+  KeyWithMaintainers,
+  GenNode,
+  NodeCreate,
+  NodeFetch,
+  NodeLookupByKey,
+  NodeExercises,
+  NodeRollback,
+}
+
+class Normalization[Nid, Cid] {
+
+  /** This class provides methods to normalize a transaction and embedded values.
+    *
+    * Informal spec: normalization is the result of serialization and deserialization.
+    *
+    * Here we take care of the following:
+    * - type information is dropped from Variant and Record values
+    * - field names are dropped from Records
+    * - values are normalized recursively
+    * - all values embedded in transaction nodes are normalized
+    * - version-specific normalization is applied to the 'byKey' fields of 'NodeFetch' and 'NodeExercises'
+    *
+    * We do not normalize the node-ids in the transaction here, but rather assume that
+    * aspect of normalization has already been performed (by the engine, or by
+    * deserialization).
+    *
+    * Eventually we would like that all aspects of normalization are achieved directly by
+    * the transaction which is constructed by the engine. When this is done, we will no
+    * longer need this separate normalization pass.
+    */
+
+  private type Val = V[Cid]
+  private type KWM = KeyWithMaintainers[Val]
+  private type Node = GenNode[Nid, Cid]
+  private type VTX = VersionedTransaction[Nid, Cid]
+
+  def normalizeTx(vtx: VTX): VTX = {
+    vtx match {
+      case VersionedTransaction(_, nodes, roots) =>
```

Comment:
similar point as below, I’d like to avoid mixing up things we do for convenience in tests with things we do on a production ledger so I’d prefer to not do the reversioning by default.

As for reusing `asVersionedTransaction`, I think you can already do this fairly easily, something like:
```
case VersionedTransaction(_, nodes, roots) =>
  val VersionedTransaction(version, _, _) = TransactionVersion.asVersionedTransaction(GenTransaction(nodes, roots))